### PR TITLE
networkd: Add ability to set MULTICAST flag on link

### LINKED ISSUE #9113: [Feature request] networkd: Add ability to set MULTICAST flag on link.
**Is your feature request related to a problem? Please describe.**
At this moment impossible to set `MULTICAST` flag on the link with `systemd-networkd`.
`MULTICAST` is disabled by default on GRE tunnels. `MULTICAST` flag is required on both sides for working with protocol 89 (ospf).

**Describe the solution you'd like**
Add ability to set `Multicast` bool value in `systemd.link` configuration.

**Describe alternatives you've considered**
`ip link set dev Tunnel0 multicast on`

### PR DESCRIPTION
**Pull request by ssahani**
Closes #9113

### CODE REVIEW

File: src/network/networkd-link.c [hunk 1]
```diff
@@ -1357,16 +1357,18 @@ static int link_set_flags(Link *link) {
...
-                return 0;
-
         r = sd_rtnl_message_new_link(link->manager->rtnl, &req, RTM_SETLINK, link->ifindex);
         if (r < 0)
                 return log_link_error_errno(link, r, "Could not allocate RTM_SETLINK message: %m");
 
         if (link->network->arp >= 0) {
                 ifi_change |= IFF_NOARP;
-                ifi_flags |= link->network->arp ? 0 : IFF_NOARP;
+                ifi_flags |= link->network->arp ? ~IFF_NOARP : IFF_NOARP;
```
Comment (yuwata) [inline src/network/networkd-link.c]: I guess this should be the following:
```
if (link->network->arp)
   ifi_flags &= ~IFF_NOARP;
else
   ifi_flags |= IFF_NOARP;
```
Or, more simply, use `SET_FLAG()` macro.

File: src/network/networkd-link.c [hunk 2]
```diff
@@ -1357,16 +1357,18 @@ static int link_set_flags(Link *link) {
...
+        }
+
+        if (link->network->multicast >= 0) {
+                ifi_change |= IFF_MULTICAST;
+                ifi_flags |= link->network->multicast ? IFF_MULTICAST : ~IFF_MULTICAST;
```
Comment (yuwata) [inline src/network/networkd-link.c]: Same here.

File: src/network/networkd-link.c [hunk 3]
```diff
@@ -1357,9 +1357,6 @@ static int link_set_flags(Link *link) {
         if (!link->network)
                 return 0;
 
-        if (link->network->arp < 0)
-                return 0;
-
```
Comment (poettering) [inline src/network/networkd-link.c]: shouldn't this rather be this now:

```c
if (link->network->arp < 0 && link->network->multicast < 0)
        return 0;
```

File: src/network/networkd-link.c [hunk 4]
```diff
@@ -1369,6 +1366,11 @@ static int link_set_flags(Link *link) {
                 ifi_flags |= link->network->arp ? 0 : IFF_NOARP;
```
Comment (poettering) [inline src/network/networkd-link.c]: event though not strictly necessary I figure this should use the SET_FLAG macro really

File: src/network/networkd-network.h [hunk 5]
```diff
@@ -213,6 +213,7 @@ struct Network {
         struct ether_addr *mac;
         uint32_t mtu;
         int arp;
+        int multicast;
```
Comment (poettering) [inline src/network/networkd-network.h:216]: code is missing to set this to -1 by default?

**Comment by ssahani** [comment]
updated thanks !

File: src/network/networkd-link.c [hunk 6]
```diff
@@ -1366,7 +1366,12 @@ static int link_set_flags(Link *link) {
 
         if (link->network->arp >= 0) {
                 ifi_change |= IFF_NOARP;
-                ifi_flags |= link->network->arp ? 0 : IFF_NOARP;
+                SET_FLAG(ifi_flags, IFF_NOARP, link->network->arp);
```
Comment (yuwata) [inline src/network/networkd-link.c]: `SET_FLAG(ifi_flags, IFF_NOARP, link->network->arp == 0);` ?

Comment (ssahani) [reply src/network/networkd-link.c]: oops right thanks :)

**Comment by ssahani** [comment]
updated

File: src/network/networkd-link.c [hunk 7]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (yuwata) [inline src/network/networkd-link.c]: spurious new line...

File: src/network/networkd-network.h (hunk 5 above)
Comment (ssahani) [reply src/network/networkd-network.h:216]: yes fixed

File: src/network/networkd-link.c (hunk 7 above)
Comment (ssahani) [reply src/network/networkd-link.c]: done

**Comment by yuwata** [comment]
lgtm.

**Pull request merged**